lib/pos_consensus.go

```diff
@@ -0,0 +1,685 @@
+		// Nothing we can do to recover here.
+		return errors.Errorf("FastHotStuffConsensus.HandleBlock: Error processing block as new tip: %v", err)
+	}
+
+	// If there are missing block hashes, then we need to fetch the missing blocks from the network
+	// and retry processing the block as a new tip. We'll request the blocks from the same peer.
+	if len(missingBlockHashes) > 0 {
+		pp.QueueMessage(&MsgDeSoGetBlocks{
+			HashList: missingBlockHashes,
+		})
+	}
+
+	return nil
+}
+
+// tryProcessBlockAsNewTip tries to apply a new tip block to both the Blockchain and FastHotStuffEventLoop data
+// structures. It wraps the ProcessBlockPoS and ProcessTipBlock functions in the Blockchain and FastHotStuffEventLoop
+// data structures, which together implement the Fast-HotStuff block handling algorithm end-to-end.
+//
+// Reference Implementation:
+// https://github.com/deso-protocol/hotstuff_pseudocode/blob/6409b51c3a9a953b383e90619076887e9cebf38d/fast_hotstuff_bls.go#L573
+func (cc *FastHotStuffConsensus) tryProcessBlockAsNewTip(block *MsgDeSoBlock) ([]*BlockHash, error) {
+	// Try to apply the block locally as the new tip of the blockchain
+	successfullyAppliedNewTip, _, missingBlockHashes, err := cc.blockchain.processBlockPoS(
+		block, // Pass in the block itself
+		cc.fastHotStuffEventLoop.GetCurrentView(), // Pass in the current view to ensure we don't process a stale block
+		true, // Make sure we verify signatures in the block
+	)
+	if err != nil {
+		return nil, errors.Errorf("Error processing block locally: %v", err)
+	}
+
+	// If the incoming block is an orphan, then there's nothing we can do. We return the missing ancestor
+	// block hashes to the caller. The caller can then fetch the missing blocks from the network and retry
+	// if needed.
+	if len(missingBlockHashes) > 0 {
+		return missingBlockHashes, nil
+	}
+
+	// At this point we know that the blockchain was mutated. Either the incoming block resulted in a new
+	// tip for the blockchain, or the incoming block was part of a fork that resulted in a change in the
+	// safe blocks
+
+	// Fetch the safe blocks that are eligible to be extended from by the next incoming tip block
+	safeBlocks, err := cc.blockchain.GetSafeBlocks()
+	if err != nil {
+		return nil, errors.Errorf("error fetching safe blocks: %v", err)
+	}
+
+	// Fetch the validator set at each safe block
+	safeBlocksWithValidators, err := cc.fetchValidatorListsForSafeBlocks(safeBlocks)
+	if err != nil {
+		return nil, errors.Errorf("error fetching validator lists for safe blocks: %v", err)
+	}
+
+	// If the block was processed successfully but was not applied as the new tip, we need up date the safe
+	// blocks in the FastHotStuffEventLoop. This is because the new block may be safe to extend even though
+	// it did not result in a new tip.
+	if !successfullyAppliedNewTip {
+		// Update the safe blocks to the FastHotStuffEventLoop
+		if err = cc.fastHotStuffEventLoop.UpdateSafeBlocks(safeBlocksWithValidators); err != nil {
```

Comment:
Hmmm... it feels like you don't need to store the safe blocks on the fc object. Rather, you can just compute them on the fly from the blockchain object every time. This would be like a *tiny* bit less efficient, but it would also be less redundant and less error-prone.

Is there a reason you want to store and update a list of safeblocks on the fc object rather than recomputing them on the fly every time?

Reply:
Ah I see... it looks like the fastHotStuffEventLoop object doesn't have a blockchain object, which is why you have to do this update manually. This makes sense, and I like the design because it makes it much easier to test the consensus code.

Reply:
Ok but I'm still confused about something. Don't you *always* need to call UpdateSafeBlocks, even if the new block was added to the tip? I'm not seeing where FastHotStuffEventLoop.safeBlocks gets updated in the case where we just attach a block to the tip, even though I think we should be adding the tip block to safeBlocks in that case.

Reply:
Sorry nevermind. It's a few lines later, and gets updated in FastHotStuffEventLoop.ProcessTipBlock().